cmds/core/cpio/cpio.go

```diff
@@ -78,20 +78,46 @@ func main() {
 			if err != nil {
 				log.Fatalf("error reading records: %v", err)
 			}
-			debug("Creating %s\n", rec)
+			debug("record name %s ino %d\n", rec.Name, rec.Info.Ino)
 
 			// A file with zero size could be a hard link to another file
-			// in the archive. The file with contents always comes first.
-			if rec.Mode&cpio.S_IFMT != cpio.S_IFDIR && rec.Info.FileSize == 0 {
-				if _, ok := inums[rec.Info.Ino]; ok {
+			// in the archive. The file with content always comes first.
+			//
+			// But we should ignore files with Ino of 0; that's an illegal value.
+			// The current most common use of this command is with u-root
+			// initramfs cpio files on Linux and Harvey.
+			// (nobody else cares about cpio any more save kernels).
+			// Those always have Ino of zero for reproducible builds.
+			// Hence doing the Ino != 0 test first saves a bit of work.
+			if rec.Info.Ino != 0 {
```

Comment:
My main concern is this breaks backwards compatibility with old cpio images created with u-root. Older cpio files made by u-root may still have files with Ino=0.

Reply:
Never mind, I see it now. It makes sense.

Reply:
But: this is just the test for a file with an ino same as a previous file. Since all inos are 0 in files generated by the u-root command, that test makes no sense.  Note that the u-root command still generates file with inos all 0.

The files are still unpacked just fine. I just tested this by running an old u-root command and unpacking with this new cpio. 

Simple test:
mkdir test
cp /etc/hosts /etc/passwd test
u-root -files test -nocmd -base /dev/null
(gnubin) cpio -ivt < /tmp/initramfs.linux_amd64.cpio 
drwxrwxr-x   0 root     root            0 Jan  1  1970 test
-rw-r--r--   0 root     root          819 Jan  1  1970 test/hosts
-rw-r--r--   0 root     root         1836 Jan  1  1970 test/passwd
7 blocks
(u-root) cpio -v i < /tmp/initramfs.linux_amd64.cpio

rminnich@a300:~/go/src/github.com/u-root/u-root/cmds/core/cpio/tp$ ls -lRi
.:
total 4
41689887 drwxrwxr-x 2 rminnich rminnich 4096 Dec 31 07:05 test

./test:
total 8
41689888 -rw-r--r-- 1 rminnich rminnich  819 Dec 31 07:05 hosts
41689889 -rw-r--r-- 1 rminnich rminnich 1836 Dec 31 07:05 passwd

The files are extracted just fine. The compare only applies to hardlinks.